File: Cpp/878_Nth_Magical_Number.cpp
class Solution {
public:
    int nthMagicalNumber(int N, int A, int B) {
        int LCM = std::lcm(A,B);
        long magical_number_count = LCM / A + LCM / B - 1;
        
        long lo = N / magical_number_count * LCM;
        long hi = lo + LCM;
        long mi = 0;

        while(true) {
            mi = (hi + lo) / 2;
            magical_number_count = mi / A + mi / B - mi / LCM;
            
            if (magical_number_count > N) hi = mi;
            else if(magical_number_count < N) lo = mi + 1;
            else break;
        }
        
        return (mi - std::min(mi % A, mi % B)) % static_cast<int>(1e9 + 7);
    }
};

File: Kotlin/878_Nth_Magical_Number.kt
class Solution {
    fun nthMagicalNumber(N: Int, A: Int, B: Int): Int {
        fun gcd(a: Long, b: Long): Long = if (b == 0L) a else gcd(b, a % b)
        fun lcm(a: Long, b: Long): Long = a / gcd(a, b) * b
        
        var lcm_value = lcm(A.toLong(), B.toLong())
        var magical_number_count = lcm_value / A + lcm_value / B - 1
        var lo = N / magical_number_count * lcm_value
        var hi = lo + lcm_value
        var mi = 0L
        
        while(true) {
            mi = (lo + hi) / 2
            magical_number_count = mi / A + mi / B - mi / lcm_value
            
            if(magical_number_count > N) hi = mi
            else if(magical_number_count < N) lo = mi + 1
            else break
        }
        
        return ((mi - Math.min(mi % A.toLong(), mi % B.toLong())) % (Math.pow(10.0, 9.0).toLong() + 7)).toInt()
    }
}